### ISSUE TITLE
How to handle uniqueness constraints?

### ISSUE THREAD

**Issue by mcls** [rom-rb/rom-rails#14]
I saw `validates_uniqueness_of` doesn't work.
I'm guessing this is because it requires access to the database during validation?

So I was wondering how (or where) these kind of "database-dependent" validations should be implemented?

**Comment by solnic** [comment]
That's why you have separate validator objects. It's not an optimal solution but I've found it to be a good fit in Rails context. Here's how you could do it *now*:

``` ruby
class NewUserValidator
  def self.call(params)
    params.valid?
    params.errors.add(:email, :taken) unless unique_email?(params.email)
  end

  def self.unique_email?(email)
    ROM.env.relations.users.where(email: email).count.zero?
  end
end
```

Please note that this is very ugly and verbose. I'm planning to add an interface for that so it will work without the need to write any custom code.

**Comment by mcls** [comment]
Ok, that makes sense :smiley:  

Now I'm confused about the `ActiveModel::Validations` being defined on the params/input object instead of the validator.

Is that because it's easier to include the `ActiveModel::Validations` module there?

**Comment by solnic** [comment]
The reason for that is to meet Rails expectations which as an object with attributes *and* errors :( This way it Just Works™ with form helpers and libraries like simple_form.

I agree, it is confusing. I would love to find a better solution to this problem that would also work with Rails assumptions. I believe it should be possible to to mix `ActiveModel::Validations` into validator module and do some tricks to make it work with params injected into validator instance. I haven't tried to do it yet so I'm just guessing :)

Please notice that we want to separate input params and validations eventually, so this would actually be inline with what we want to achieve.

**Comment by mcls** [comment]
Ok, so I got this to work after accounting for a subtle issue.

In rails `#valid?` clear the errors first (see https://github.com/rails/rails/blob/v4.1.7.1/activemodel/lib/active_model/validations.rb#L314-L320). 

So in order to add errors in the validator object I had to check for `params.errors.blank?` instead of [`params.valid?`](https://github.com/rom-rb/rom-rails/blob/master/lib/rom/model.rb#L92) and then rails the `ROM::Model::ValidationError`.

Basically the snippet became:

```
class NewUserValidator
  def self.call(params)
    params.valid?
    params.errors.add(:email, :taken) unless unique_email?(params.email)
    unless params.errors.blank?
      rails ROM::Model::ValidationError.new(params)
    end
  end

  def self.unique_email?(email)
    ROM.env.relations.users.where(email: email).count.zero?
  end
end
```

**Comment by solnic** [comment]
@mcls oh sorry, totally forgot about raising the error part :) I reported a related issue which describes potential improvement => #15
